packages/babel-parser/src/parser/expression.js

```diff
@@ -539,37 +524,31 @@ export default class ExpressionParser extends LValParser {
     );
   }
 
-  // Helper function for `parseExprOpRightExpr` for the Hack-pipe operator
-  // (and the Hack-style smart-mix pipe operator).
-
-  parseHackPipeBody(op: TokenType, prec: number): N.Expression {
-    // If the following expression is invalidly a `yield` expression,
-    // then throw a human-friendly error.
-    // A `yield` expression in a generator context (i.e., a [Yield] production)
-    // starts a YieldExpression.
-    // Outside of a generator context, any `yield` as a pipe body
-    // is considered simply an identifier.
-    // This error is checked here, before actually parsing the body expression,
-    // because `yield`’s “not allowed as identifier in generator” error
-    // would otherwise have immediately
-    // occur before the pipe body is fully parsed.
-    // (Unparenthesized assignment expressions are handled
-    // in `parseMaybeAssign`,
-    // and unparenthesized `=>` expressions are handled
-    // in `checkHackPipeBodyEarlyErrors`.)
-    const bodyIsInGeneratorContext = this.prodParam.hasYield;
-    const bodyIsYieldExpression =
-      bodyIsInGeneratorContext && this.isContextual("yield");
-
-    if (bodyIsYieldExpression) {
-      throw this.raise(
-        this.state.start,
-        Errors.PipeBodyIsTighter,
-        this.state.value,
+  parseHackPipeBody(): N.Expression {
+    const { start } = this.state;
+
+    const body = this.parseMaybeAssign();
+
+    const invalidBodies = {
+      ArrowFunctionExpression: "arrow function",
+      AssignmentExpression: "assignment",
+      ConditionalExpression: "conditional",
+      YieldExpression: "yield",
+    };
+
+    if (hasOwn(invalidBodies, body.type) && !body.extra?.parenthesized) {
```

Comment:
We can transform `invalidBodies` to a map so it comes with `.has()`, and it's also faster than objects.